src/Streamly/Mem/Array.hs

```diff
@@ -359,27 +453,55 @@ writeIndex arr i a = do
             liftIO $ withForeignPtr (aStart arr) $ \p ->
                 pokeElemOff p i a
 
-{-
+
 -- | @writeSlice arr i count stream@ writes a stream to the array @arr@
 -- starting at index @i@ and writing up to @count@ elements in the forward
 -- direction ending at the index @i + count - 1@.
 --
 -- @since 0.7.0
 {-# INLINE writeSlice #-}
-writeSlice :: (IsStream t, Monad m, Storable a)
+writeSlice :: (IsStream t, MonadIO m, Storable a)
```

Comment:
Comments in readSlice apply here as well. Avoid `writeIndex` as performs bounds checks every time.